drivers/gpio/gpio_keys_zephyr.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) 2022 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio_keys.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(zephyr_gpio_keys, LOG_LEVEL_DBG);
+
+#define DT_DRV_COMPAT zephyr_gpio_keys
+
+struct gpio_keys_config {
+	/** GPIO specification from devicetree */
+	struct gpio_dt_spec spec;
+	/** Debounce interval in milliseconds from devicetree */
+	uint32_t debounce_interval_ms;
+	/** Zephyr code from devicetree */
+	uint32_t zephyr_code;
+};
+
+struct gpio_keys_pin_data {
+	const struct device *dev;
+	struct gpio_keys_callback cb_data;
+	struct k_work_delayable work;
+	int8_t pin_state;
+};
+
+struct gpio_keys_data {
+	gpio_keys_callback_handler_t callback;
+	int num_keys;
+	struct gpio_keys_pin_data *keys;
+};
+
+/**
+ * Handle debounced gpio pin state.
+ */
+static void gpio_keys_change_deferred(struct k_work *work)
+{
+	struct gpio_keys_pin_data *pin_data = CONTAINER_OF(work, struct gpio_keys_pin_data, work);
+	const struct device *dev = pin_data->dev;
+	struct gpio_keys_data *data = dev->data;
+	int key_index = pin_data - &data->keys[0];
+	const struct gpio_keys_config *cfg = &((struct gpio_keys_config *)dev->config)[key_index];
```

Comment:
I think you don't need the casting here, `config` is `void *`.

Reply:
I'll clean this up.  I was casting it so the array indexing operator knows how to iterate the array.

Reply:
Oh good point, did not realize that, thanks for explaining.